### ISSUE TITLE
Blazor Stream Rendering not working on IIS Express

### ISSUE THREAD

**Issue by brackers17** [dotnet/aspnetcore#52323]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Describe the bug

When going to the weather forecasts page using IIS Express it waits for the 5 second delay and then shows the page.

The URL correctly changes but the page content doesn't until the forecast data is loaded.

![image](https://github.com/dotnet/aspnetcore/assets/45168007/37796120-ddc1-4359-a16b-6250f89153f6)

### Expected Behavior

It should show the weather heading with the text "This component demonstrates showing data." and show "loading..." until the 5 second delay has finished then show the forecast data.

The below print screen is what we expect to happen. This was created by refreshing the forecasts page.

![image](https://github.com/dotnet/aspnetcore/assets/45168007/729247b8-fb3e-446c-b436-61208cdea815)

### Steps To Reproduce

1. Scaffold the new Blazor web app template using settings shown here: ![image](https://github.com/dotnet/aspnetcore/assets/45168007/c71f2496-d368-4928-9aac-cf41bcea0202)
2. Change the delay in weather.razor to 5000 to make the issue more obvious
3. Run the app using IIS Express in Visual Studio
4. Click on the Weather button

### .NET Version

8.0.100

### Anything else?

Visual Studio 2022 version 17.8.1

Dotnet Info:

.NET SDK:
 Version:           8.0.100repos\Test Projects\BlazorDotNet8\BlazorDotNet8> ^Ctnet --version
 Commit:            57efcf1350os\Test Projects\BlazorDotNet8\BlazorDotNet8> dotnet --info
 Workload version:  8.0.100-manifests.8d38d0cc

Runtime Environment:
 OS Name:     Windows
 OS Version:  10.0.19045
 OS Platform: Windows
 RID:         win-x64
 Base Path:   C:\Program Files\dotnet\sdk\8.0.100\

.NET workloads installed:
 Workload version: 8.0.100-manifests.8d38d0cc
 [android]                                                                                                   oadManifest.json
   Install Type:              Msi4316.72
   Manifest Version:    34.0.43/8.0.100
 [maui-windows]h:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.android\34.0.43\Worklnifest.json
   Install Type:              Msi4316.72
   Manifest Version:    8.0.3/8.0.100
 [maccatalyst]th:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maui\8.0.3\WorkloadMa\WorkloadManifest.json
   Install Type:              Msi4316.72
   Manifest Version:    17.0.8478/8.0.100
 [ios]ifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.maccatalyst\17.0.8478dManifest.json
   Install Type:              Msi4316.72
   Manifest Version:    17.0.8478/8.0.100
   Manifest Path:       C:\Program Files\dotnet\sdk-manifests\8.0.100\microsoft.net.sdk.ios\17.0.8478\Workload
Host:
  Version:      8.0.0
  Architecture: x64
  Commit:       5535e31a71

.NET SDKs installed:
  1.1.8 [C:\Program Files\dotnet\sdk]
  1.1.9 [C:\Program Files\dotnet\sdk]
  1.1.10 [C:\Program Files\dotnet\sdk]
  1.1.11 [C:\Program Files\dotnet\sdk]
  1.1.14 [C:\Program Files\dotnet\sdk]
  2.1.101 [C:\Program Files\dotnet\sdk]
  2.1.102 [C:\Program Files\dotnet\sdk]
  2.1.201 [C:\Program Files\dotnet\sdk]
  2.1.202 [C:\Program Files\dotnet\sdk]
  2.1.300 [C:\Program Files\dotnet\sdk]
  2.1.301 [C:\Program Files\dotnet\sdk]
  2.1.400 [C:\Program Files\dotnet\sdk]
  2.1.402 [C:\Program Files\dotnet\sdk]
  2.1.403 [C:\Program Files\dotnet\sdk]
  2.1.526 [C:\Program Files\dotnet\sdk]
  2.2.402 [C:\Program Files\dotnet\sdk]
  3.1.426 [C:\Program Files\dotnet\sdk]
  5.0.410 [C:\Program Files\dotnet\sdk]
  8.0.100 [C:\Program Files\dotnet\sdk]

.NET runtimes installed:
  Microsoft.AspNetCore.All 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.All 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.All]
  Microsoft.AspNetCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.AspNetCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.AspNetCore.App]
  Microsoft.NETCore.App 1.0.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.11 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.12 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.0.16 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.8 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.10 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 1.1.13 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.6 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.0.9 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.1 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.2 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.4 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.5 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.26 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.1.30 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 2.2.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.NETCore.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.NETCore.App]
  Microsoft.WindowsDesktop.App 3.1.27 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 3.1.32 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.7 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.15 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 6.0.25 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 7.0.14 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]
  Microsoft.WindowsDesktop.App 8.0.0 [C:\Program Files\dotnet\shared\Microsoft.WindowsDesktop.App]

Other architectures found:
  x86   [C:\Program Files (x86)\dotnet]
    registered at [HKLM\SOFTWARE\dotnet\Setup\InstalledVersions\x86\InstallLocation]

Environment variables:
  Not set

global.json file:
  Not found

**Comment by surayya-MS** [comment]
Thanks for contacting us @brackers17 !
It seems that you are buffering the response in IIS, in order to benefit from streaming rendering you need to disable buffering at the IIS level.

**Comment by surayya-MS** [comment]
Setting this in your IIS config might help avoid the buffering (and at that point streaming rendering should work).
```
<configuration>
<system.webServer>
<handlers>
<add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" responseBufferLimit="0" />
</system.webServer>
</configuration>
```

**Comment by brackers17** [comment]
@surayya-MS I have tried updating the applicationHost.config file for IIS express with the above but I'm afraid it hasn't made a difference. 

Given that when you refresh the page it correctly comes back with "Loading..." does this not indicate that the IIS config was already correct?

**Comment by ZeraGames** [comment]
I also experience the same problem but not just for "IIS Express" but also with "https" and "http" modes.
Just upgraded to VS17.8.1 and .net8

**Comment by brackers17** [comment]
@amcasey No, its working fine on https for me.

**Comment by amcasey** [comment]
Yep, the same thing happens on my box.  Thanks for the detailed steps!

**Comment by amcasey** [comment]
@ZeraGames Can you please file a separate issue.  If you're seeing problems that aren't confined to IIS, it may be unrelated (and we can always merge them again later).

**Comment by amcasey** [comment]
The trace level logs for Kestrel and IIS are basically the same - just some server specific messages that appear in one but not the other.

**Comment by ahmetsarikaya0696** [comment]
I have the same problem. I tried with IIS and also Kestrel. Both have the same issue.

**Comment by ahmetsarikaya0696** [comment]
> Setting this in your IIS config might help avoid the buffering (and at that point streaming rendering should work).
> 
> ```
> <configuration>
> <system.webServer>
> <handlers>
> <add name="aspNetCore" path="*" verb="*" modules="AspNetCoreModuleV2" resourceType="Unspecified" responseBufferLimit="0" />
> </system.webServer>
> </configuration>
> ```

I checked my web.config file and it's already the same like what you recommend.

**Comment by BrennanConroy** [comment]
Looks like the stream rendering is being compressed by default in IISExpress which makes the streaming part stop working.

You can fix this by modifying the `<repo root>/.vs/<app name>/config/applicationhost.config` file:
```
httpCompression directory="%TEMP%">
      <scheme name="gzip" dll="%IIS_BIN%\gzip.dll" />
      <dynamicTypes>
        <add mimeType="text/*" enabled="false" /> <!-- this is true by default -->
```

This would also probably occur if you install the dynamic compression module in IIS and enabled it for your app.

**Comment by ahmetsarikaya0696** [comment]
> @ahmetsarikaya0696 If you're seeing problems in Kestrel too, that's probably a different issue.  Can you please file one and describe your setup and what you're seeing?

I didn't change anything. Just created a project with sample pages. Only change I did is disabling nullable. If you ask for a specific folder or configuration I will be providing that.

**Comment by alexdulemba** [comment]
I'm having similar issues with the stream rendering feature. When I run my application in `https` then it works as expected, but it does not work with `http` or `IIS Express` profiles

**Comment by javiercn** [comment]
This seems to happen in the following circumstances:
* Streaming rendering is enabled.
* App is using enhanced navigation.
* Response is compressed by IIS (Express).

It works fine if not using enhanced nav (works on a regular server refresh).

With this in mind, I don't think there's anything that we can do on our side, other than documenting this. It's not possible to disable compression from JS, and it seems that either chrome or something else on the stack is buffering the entire response.

The recommendation here is to apply @BrennanConroy suggestion as you shouldn't be compressing dynamic HTML responses coming from Blazor (or any other SSR framework for that matter) by default as that opens you to BREACH attacks.

**Comment by javiercn** [comment]
@davidfowl that's good to know thanks!

That seems to do the trick!

**Comment by davidfowl** [comment]
We should attempt to patch this in .NET 8

**Comment by javiercn** [comment]
@davidfowl I've already put it in the patch milestone 😄

**Comment by andystumpp** [comment]
Interesting. I'm seeing this issue only when deployed to App Service, not locally.
Fixed via adding _web.config_ with the following:
```
<configuration>
	<system.webServer>
		<urlCompression doStaticCompression="false" doDynamicCompression="false" />
	</system.webServer>
</configuration>
```

**Comment by andoband** [comment]
I was seeing this issue mostly when streaming enhanced form posts.

I turned off compression per the suggestions above for local dev, and intermediate responses streamed back worked intermittently.  Then I switched from IIS to Kestral for local development and the streaming now works well.

Although, it doesn't work when deploying to App Service (Windows)...which has IIS involved.  Turning off compression got the App Service to work intermittently.  But it'd be nice to have it work consistently, and I also don't like turning off compression.  

Ideally, there would be a way to turn off IIS compression/buffering just for enhanced form posts (or maybe Enhanced Nav in general).  Maybe there is already?
